### PR TITLE
Jon sub

### PR DESCRIPTION
**Pull request by richjoregonstate**
Everything from #98 is in and tested this time around 👍

### CODE REVIEW

**Comment by richjoregonstate** [comment]
Also, code coverage is up to 95% 

image

File: src/main/java/assets/game.js [hunk 1]
```diff
@@ -113,11 +113,14 @@ function cellClick() {
     let col = String.fromCharCode(this.cellIndex + 65);
     let useSonar = document.getElementById("use_sonar").checked;
     if (isSetup) {
+        if(shipType === "SUB" && document.getElementById("place_submarine_underwater").checked){// Flag as an underwater sub
+            shipType = "SUBB";
+        }
```
Comment (yorkal) [inline src/main/java/assets/game.js]: Makes sense.  Glad you created a separation between above and bellow water submarine.

File: src/main/java/assets/game.js [hunk 2]
```diff
@@ -176,6 +184,39 @@ function place(size) {
...
+                // ship is over the edge; let the back end deal with it
+                sub = false;
+                size = 5;
+                return;
+            }
+            cell.classList.toggle("placed");
+
+            sub = false;
+            size = 5;
+        }
```
Comment (yorkal) [inline src/main/java/assets/game.js]: Looks good.  Uses the same ideas as the rest of the place ship function.  You handle the sub being placed over the edge of the board well.

File: src/main/java/cs361/battleships/models/Board.java [hunk 3]
```diff
@@ -27,16 +28,23 @@ public Board() {
...
+		}
 		placedShip.place(y, x, isVertical);
-		if (ships.stream().anyMatch(s -> s.overlaps(placedShip))) {
-			return false;
+		for(int i=0;i<ships.size();i++){// Rewritten to check underwater status.
+			if(ships.get(i).overlaps(placedShip)){// If there is an overlap
+				if(ships.get(i).isUnderwater() == placedShip.isUnderwater()) {// If they are both under/above water error out
+					return false;
+				}//Otherwise we are good
+			}
```
Comment (yorkal) [inline src/main/java/cs361/battleships/models/Board.java]: Good check to make sure two ships in the same position on board and same level (submerged/surface) can't be placed.

File: src/main/java/cs361/battleships/models/Board.java [hunk 4]
```diff
@@ -161,6 +169,27 @@ else if (capNumB == 0){
...
+					attacks.add(attackResult);
+					tmpS = hitShip.getOccupiedSquares().get(4);
+					attackResult = hitShip.attack(tmpS.getRow(), tmpS.getColumn());
+				}
+				else if (capNumS == 0){
+					capNumS++;
+					attackResult.setResult(AtackStatus.CAPHIT);
+					hitShip.getCCM().nohit();
+				}
+			}
```
Comment (yorkal) [inline src/main/java/cs361/battleships/models/Board.java]: This follows what we did for the other types of ships.  Looks good and keeps the code consistent.

File: src/main/java/cs361/battleships/models/Ship.java [hunk 5]
```diff
@@ -25,28 +26,63 @@ public Ship(String kind) {
...
 				break;
+			case "SUB":
+				this.underwater = false;
+				size = 5;
+				break;
+			case "SUBB":
+				this.underwater = true;
+				this.kind = "SUB";
+				size = 5;
+				break;
```
Comment (yorkal) [inline src/main/java/cs361/battleships/models/Ship.java]: Good OO design.  Keeps whether a ship is underwater or not in the Ship class.

File: src/main/java/cs361/battleships/models/Ship.java [hunk 6]
```diff
@@ -25,28 +26,63 @@ public Ship(String kind) {
...
+			default:
+				for (int i = 0; i < size; i++) {
+					if (isVertical) {
+						occupiedSquares.add(new Square(row + i, col));
+					} else {
+						occupiedSquares.add(new Square(row, (char) (col + i)));
+					}
+				}
+				break;
```
Comment (yorkal) [inline src/main/java/cs361/battleships/models/Ship.java]: This is important.  Handles the new shape of the sub well while also keeping functionality for other ship types.  Good job.

File: src/main/java/views/ApplicationController/index.ftl.html [hunk 7]
```diff
@@ -53,6 +53,8 @@ <h1 style="color: #ffffff; text-align: center;">The Battleship game</h1>
             <li><button id="place_minesweeper">Place Minesweeper</button></li>
             <li><button id="place_destroyer">Place Destroyer</button></li>
             <li><button id="place_battleship">Place Battleship</button></li>
+            <li><button id="place_sub">Place Submarine</button></li>
+            <li><input type="checkbox" id="place_submarine_underwater">Place Submarine underwater</input></li>
```
Comment (yorkal) [inline src/main/java/views/ApplicationController/index.ftl.html]: Good job remembering to add UI elements.

File: src/test/java/cs361/battleships/models/BoardTest.java [hunk 8]
```diff
@@ -186,4 +187,47 @@ public void testAtackGame(){
...
+        b.attack(4,'A',false);
+        b.attack(5,'C',false);
+        total = 0;
+        for(int i=0;i < b.getAtacks().size();i++){
+            if(b.getAtacks().get(i).getResult().equals(AtackStatus.SUNK) || b.getAtacks().get(i).getResult().equals(AtackStatus.SURRENDER)){
+                total++;
+            }
+        }
+        assertEquals(4,total);
+    }
```
Comment (yorkal) [inline src/test/java/cs361/battleships/models/BoardTest.java]: Tests look good.  You check the captain's quarters functionality and the underwater functionality which is pretty much all that Is needed for the sub.

**Pull request merged**